oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java

```diff
@@ -212,12 +214,19 @@ public static SecretKeyJwtDecoderBuilder withSecretKey(SecretKey secretKey) {
 		private String jwkSetUri;
 		private Set<SignatureAlgorithm> signatureAlgorithms = new HashSet<>();
 		private RestOperations restOperations = new RestTemplate();
+		private JWKSetCache jwkSetCache = new DefaultJWKSetCache();
 
 		private JwkSetUriJwtDecoderBuilder(String jwkSetUri) {
 			Assert.hasText(jwkSetUri, "jwkSetUri cannot be empty");
 			this.jwkSetUri = jwkSetUri;
 		}
 
+		public JwkSetUriJwtDecoderBuilder jwkSetCache(JWKSetCache jwkSetCache) {
```

Comment:
I'd prefer exposing an interface that is more powerful than the `JWKSetCache` interface. If an application really wants to supply a `JWKSetCache`, then it's quite simple to construct a `DefaultJWTProcessor` directly and pass that to the `NimbusJwtDecoder` constructor.

What if this took a `org.springframework.cache.Cache` instead? Then this builder could internally wrap that in a `CacheJWKSetCache` implementation, similar to the design of `RestOperationsResourceRetreiver`.

The nice thing about doing that is that an application can then use whatever caching mechanism they wish, be it Caffeine, Hazelcast, or something else. 

Another nice this is that using `Cache` allows the entries to be key-value based which is better suited for a multi-tenant resource server that is caching keys from multiple issuers.

Reply:
@jzheaux Sry, but how should it help to multi-tenant resource server? I mean in multi-tenant environment we will have a separate `AuthenticationManager` for each tenant, that means unique `jwks-uri` for each tenant -> so unique `jwkSetCache` for each tenant, there is no way multiple keys will be stored with the current implementation

Reply:
Good question, @20fps. I guess it comes down to what parts of your infra are multi-tenant. If your cache is multi-tenant, then each `AuthenticationManager` could have a `JwtDecoder` that shares the same cache - each JWK Set keyed, for example, by that tenant's JWK Set Uri. 

If each tenant has its own cache, then that point is moot. My main point is that by introducing a key to the cache grants additional flexibility - multi-tenancy is one way that flexibility could be leveraged.

Reply:
@jzheaux still don't see it can happen since we have the interface method `jwkSetCache#get`, it means we cannot pass any key there, so the only way spring Cache can be used is the hardcoded key only with only one element. Do you have other ideas how to resolve it?

Reply:
The possible way could be to allow passing the key from outside, then the method should be like jwkSetCache(Object key, Cache jwkSetCache), so we can have this data inside inner implementation and do something like `get() { return jwkSetCache.get(key); }`, what do you think? Just don't see another way without reimplementing stuff like `RemoteJWKSet`